### DISCUSSION TITLE
Feature Request: Add Support for 3-Finger Touchpad Gesture to Navigate Active Spaces

### DISCUSSION THREAD

**Discussion by joeonit** [nikitabobko/AeroSpace#1003] [feature-ideas]
Please add support for using the 3-finger touchpad gesture to navigate through active spaces. This would make it easier for users who are mainly reading or viewing content, allowing seamless space switching without relying on keyboard shortcuts.
Thank you!

**Comment by hwamil** [comment]
https://nikitabobko.github.io/AeroSpace/goodness#use-trackpad-gestures-to-switch-workspaces

**Comment by joeonit** [comment]
> https://nikitabobko.github.io/AeroSpace/goodness#use-trackpad-gestures-to-switch-workspaces

Isn’t there any workaround for this, as BetterTouchTool is a paid product?

**Comment by jakenvac** [comment]
You can research other tools to create custom trackpad gestures. Here's one such tool, however it's also a paid product.

https://multitouch.app/

I would recommend paying for tools that provide value to you.

It's a lot more work, but you could also do this with [hammerspoon](https://github.com/Hammerspoon/hammerspoon). Looks like there's a package for hammerspoon to make swipe detection much easier: https://github.com/mogenson/Swipe.spoon

**Comment by jakenvac** [comment]
@joeonit Here's a working hammerspoon script, using the 'swipe.spoon' module from my previous comment. It works great - it's a four finger swipe to change workspaces.

You can change it to a 3 (or n) finger swipe by changing the first argument to `Swipe:start`. Make sure you disable this gesture in the macos trackpad settings to avoid conflicts.

```lua
Swipe = hs.loadSpoon("Swipe")

local AEROSPACE = "/opt/homebrew/bin/aerospace"

function aerospaceExec(cmd)
  os.execute("nohup " .. AEROSPACE .. " " .. cmd .. " &")
end

-- use four finger swipe to switch workspace
local current_id, threshold
Swipe:start(4, function(direction, distance, id)
  if id == current_id then
    if distance > threshold then
      threshold = math.huge -- only trigger once per swipe
      if direction == "left" then
        aerospaceExec("workspace --wrap-around prev")
      elseif direction == "right" then
        aerospaceExec("workspace --wrap-around next")
      end
    end
  else
    current_id = id
    threshold = 0.1 -- swipe distance > 10% of trackpad
  end
end)

```

I'll submit a PR to add this to the goodness page later.

**Comment by monsterhxw** [comment]
@jakenvac 

Thank you for your solution. 
Based on your Hammerspoon solution, I made some optimizations. Hope this can help others who find this later.

```lua
-- https://github.com/mogenson/Swipe.spoon
Swipe = hs.loadSpoon("Swipe")

local config = {
  fingers = 4,
  -- 0.1 = swipe distance > 10% of trackpad
  threshold = 0.08,
  showAlert = true,
  alertDuration = 0.4
}

local AEROSPACE = "/usr/local/bin/aerospace"
function aerospaceExec(cmd)
  local command = string.format('%s list-workspaces --monitor mouse --visible | xargs %s workspace && %s workspace %s', AEROSPACE, AEROSPACE, AEROSPACE, cmd)

  hs.execute(command)
  
  if config.showAlert then
    hs.alert.show("AeroSpace: " .. cmd, config.alertDuration)
  end
end

local current_id, threshold
-- use 4-fingers swipe to switch workspace
Swipe:start(config.fingers, function(direction, distance, id)
  if id == current_id then
    if distance > threshold then
      -- only trigger once per swipe
      threshold = math.huge
      if direction == "left" then
        aerospaceExec("next")
      elseif direction == "right" then 
        aerospaceExec("prev")
      end
    end
  else
    current_id = id
    threshold = config.threshold
  end
end)
```

- https://github.com/monsterhxw/dotfiles/blob/main/hammerspoon/scripts/aerospace-trackpad-gesture.lua

**Comment by mayurankv** [comment]
Is this intended to be added as an option? (No worries if not, just curious!)

**Comment by laHeud** [comment]
When I used `os.execute()` or `hs.execute()` to trigger AeroSpace commands, Hammerspoon froze/lagged because these functions are **blocking** - they wait for the command to complete before continuing.

I found that using `hs.task` instead works perfectly. It's **asynchronous** and non-blocking:

```lua
local AEROSPACE = "/opt/homebrew/bin/aerospace"

function aerospaceExec(cmd)
  local fullCmd = string.format('/opt/homebrew/bin/aerospace list-workspaces --monitor mouse --visible | xargs /opt/homebrew/bin/aerospace workspace && /opt/homebrew/bin/aerospace workspace %s', cmd)
  
  -- Asynchronous execution - no lag!
  local task = hs.task.new("/bin/bash", nil, {"-c", fullCmd})
  task:start()
  
  hs.alert.show("AeroSpace: " .. cmd, 0.4)
end
```

**Reply by jakenvac** [reply]
This is why I was using `nohup .. &` in the command - it detaches it from the parent process to avoid blocking and hangups (`SIGHUP`). I'm glad you've found an alternative that also works for you, though.

**Comment by LunaVoid** [comment]
Hi, this is a major reason why i haven't switched to aerospace. I just found this repo (below) and I'm going to try it out. I wondered if anyone else has use it? Or is hammerspoon still the best solution?

https://github.com/acsandmann/aerospace-swipe

**Reply by laHeud** [reply]
I tried to use it but it didn't work. IMO hammerspoon is the best free solution.

**Comment by iandol** [comment]
For completeness, the BetterTouchTool solution to this:

<img width="2574" height="550" alt="Screenshot_2025-11-15-06-33-19" src="https://github.com/user-attachments/assets/03326deb-47e4-457d-8cd2-3d8e59d68e64" />

This allows the same swipe to work whether Aerospace is running or not, as it checks first and falls back to the macOS default if not. BTT is great.

**Comment by neilyhy** [comment]
https://github.com/nikitabobko/AeroSpace/pull/1684 actually broke the instructions on https://nikitabobko.github.io/AeroSpace/goodies#use-trackpad-gestures-to-switch-workspaces
at-least on an old version of BTT (4.389)
totally fine after adding --no-stdin
```
/opt/homebrew/bin/aerospace workspace "$(/opt/homebrew/bin/aerospace list-workspaces --monitor mouse --visible)" && 
/opt/homebrew/bin/aerospace workspace --no-stdin next
```

**Comment by jomatsu** [comment]
I built exactly this as a standalone companion app after fighting the same problem — sharing since this thread is where I would have looked for it:

**AeroKit** ships native-feeling three-finger workspace swiping for AeroSpace (continuous finger tracking with a workspace-strip HUD, natural direction, wrap-around, skip-empty). Free & MIT: https://github.com/jomatsu/aerokit — announcement with a demo in #2187

https://github.com/user-attachments/assets/efdee74c-3e46-4cb1-b6b4-e46ba6c1a599

```sh
brew install --cask jomatsu/tap/aerokit
```

Two notes from experience if you try it:

- set "Swipe between full-screen applications" to four fingers or off, and disable any Hammerspoon/BTT swipe handler first — two handlers reacting to the same physical gesture switch twice, which looks like random breakage
- re the Hammerspoon blocking issue discussed above: AeroKit talks to AeroSpace over its unix socket (~1–4 ms per call) off the main thread, so gestures never freeze anything

**Reply by HosseyNJF** [reply]
Was skeptical because the message was AI generated, but the app is really solid, does the job, and looks good! Check it out

**Comment by pszypowicz** [comment]
Another option for this thread: I built [Cyclist](https://github.com/nikitabobko/AeroSpace/discussions/2191), which drives workspace switching with the actual system Spaces swipe - the gesture stays enabled in System Settings, and Cyclist consumes it from the WindowServer's gesture stream before the Dock starts the animated transition. It also splices native fullscreen Spaces into the same ring as the workspaces, which is the part none of the tools above cover. One swipe is one discrete step, no continuous tracking. macOS 26 only. It's part of a Cmd+Tab replacement, but the switcher is optional - a Settings toggle leaves just the swipe and keyboard chain navigation. Details and a demo in the linked announcement.
